File: examples/php/09_probabilistic.php
<?php
require 'vendor/autoload.php';

$client = new Predis\Client([
    'scheme' => 'tcp',
    'host'   => '127.0.0.1',
    'port'   => 6379,
]);

echo "--- 09 God Tier Probabilistic Data Structures ---\n";

// 1. HyperLogLog
$client->executeRaw(['PFADD', 'hll_php', 'p1', 'p2']);
$count = $client->executeRaw(['PFCOUNT', 'hll_php']);
echo "[HLL] Count: $count\n";

// 2. Cuckoo Filter
$client->executeRaw(['CF.ADD', 'cf_php', 'test']);
$exists = $client->executeRaw(['CF.EXISTS', 'cf_php', 'test']);
echo "[Cuckoo] Exists: $exists\n";

// 3. Count-Min Sketch
$client->executeRaw(['CMS.INCRBY', 'cms_php', 'page_view', 5]);
$val = $client->executeRaw(['CMS.QUERY', 'cms_php', 'page_view']);
echo "[CMS] Count: $val\n";

// 4. Top-K
$client->executeRaw(['TOPK.ADD', 'topk_php', 'dog', 'cat', 'dog']);
$list = $client->executeRaw(['TOPK.LIST', 'topk_php']);
echo "[TopK] List: " . json_encode($list) . "\n";

// 5. t-digest
$client->executeRaw(['TDIGEST.ADD', 'td_php', 12.5]);
$res = $client->executeRaw(['TDIGEST.QUANTILE', 'td_php', 0.5]);
echo "[t-digest] Median: $res\n";
?>

File: examples/vbnet/09_Probabilistic.vb
Imports StackExchange.Redis
Imports System

Module Program
    Sub Main()
        Dim redis = ConnectionMultiplexer.Connect("localhost")
        Dim db = redis.GetDatabase()

        Console.WriteLine("--- 09 God Tier Probabilistic Data Structures ---")

        ' 1. HyperLogLog
        db.Execute("PFADD", "hll_vb", "v1", "v2")
        Dim count = db.Execute("PFCOUNT", "hll_vb")
        Console.WriteLine($"[HLL] Count: {count}")

        ' 2. Cuckoo Filter
        db.Execute("CF.ADD", "cf_vb", "itemA")
        Dim exists = db.Execute("CF.EXISTS", "cf_vb", "itemA")
        Console.WriteLine($"[Cuckoo] Exists: {exists}")

        ' 3. Count-Min Sketch
        db.Execute("CMS.INCRBY", "cms_vb", "tag1", 10)
        Dim val = db.Execute("CMS.QUERY", "cms_vb", "tag1")
        Console.WriteLine($"[CMS] Value: {val}")

        ' 4. Top-K
        db.Execute("TOPK.ADD", "topk_vb", "apple", "banana", "apple")
        Dim list = db.Execute("TOPK.LIST", "topk_vb")
        Console.WriteLine($"[TopK] List: {list}")

        ' 5. t-digest
        db.Execute("TDIGEST.ADD", "td_vb", 42.5)
        Dim res = db.Execute("TDIGEST.QUANTILE", "td_vb", 0.5)
        Console.WriteLine($"[t-digest] Median: {res}")
    End Sub
End Module
